hip/test/components/cooperative_groups.hip.cpp

```diff
@@ -0,0 +1,204 @@
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+// TODO remove when the HIP includes are fixed
+#include <hip/hip_runtime.h>
+
+
+#include "hip/components/cooperative_groups.hip.hpp"
+
+
+#include <memory>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/executor.hpp>
+
+
+namespace {
+
+
+using namespace gko::kernels::hip;
+using namespace gko::kernels::hip::group;
+
+
+class CooperativeGroups : public ::testing::Test {
+protected:
+    CooperativeGroups()
+        : ref(gko::ReferenceExecutor::create()),
+          hip(gko::HipExecutor::create(0, ref)),
+          result(ref, 1),
+          dresult(hip)
+    {
+        *result.get_data() = true;
+        dresult = result;
+    }
+
+    template <typename Kernel>
+    void test(Kernel kernel)
+    {
+        hipLaunchKernelGGL(HIP_KERNEL_NAME(kernel), dim3(1),
+                           dim3(config::warp_size), 0, 0, dresult.get_data());
+        result = dresult;
+        auto success = *result.get_const_data();
+
+        ASSERT_TRUE(success);
+    }
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<gko::HipExecutor> hip;
+    gko::Array<bool> result;
+    gko::Array<bool> dresult;
+};
+
+
+constexpr static int subwarp_size = config::warp_size / 4;
+
+
+__device__ void test_assert(bool *success, bool partial)
+{
+    auto warp = tiled_partition<config::warp_size>(this_thread_block());
+    auto full = warp.all(partial);
+    if (threadIdx.x == 0) {
+        *success &= full;
+    }
+}
+
+
+__global__ void cg_shuffle(bool *s)
+{
+    auto group = tiled_partition<config::warp_size>(this_thread_block());
+    auto i = int(group.thread_rank());
+    test_assert(s, group.shfl_up(i, 1) == max(0, i - 1));
+    test_assert(s, group.shfl_down(i, 1) == min(i + 1, config::warp_size - 1));
+    test_assert(s, group.shfl(i, 0) == 0);
```

Comment:
Could you add the complex<double> test for the extend_shuffle? 
It need a lot of threads to test it. (in #439, it works on double with extended shuffle (disabling the official ones) 
It would be nice to check whether it also works on complex double as expected